### PR TITLE
opt: add capability trimming pass (not default).

### PR DESCRIPTION
**Pull request by Keenuts**
This commit adds a new optimization which tries to remove unnecessary capabilities from a SPIR-V module.

When compiling a SPIR-V module, you may have some dead-code using features gated by a capability.
DCE will remove this code, but the capability will remain. This means your module would still require some capability, even if it doesn't require it. Calling this pass on your module would remove obsolete capabilities.

This pass wouldn't be enabled by default, and would only be usable from the API (at least for now).

NOTE: this commit only adds the basic skeleton/structure and supports the ShaderClockKHR capability as initial example.

### CODE REVIEW

File: source/opt/trim_capabilities_pass.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+      continue;
+    }
+
+    if (trimmed_extensions.count(extension) != 0) {
+      instructions_to_kill.push_back(&instruction);
+    }
+  }
+
+  // Killing an instruction invalidates the instruction iterator. We must gather
+  // & kill in 2 steps.
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:123]: Use the full words "and" and "two".

File: source/opt/trim_capabilities_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+  return required_capabilities;
+}
+
+// Utility function to, given a list of capabilities, retrieve the associated
+// extensions.
+template <typename Container>
+static ExtensionSet GetExtensionsForCapabilities(
+    const Container& capabilities) {
+  ExtensionSet output;
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:90]: There is an enum for all extensions (https://github.com/KhronosGroup/SPIRV-Tools/blob/main/source/extensions.h).

File: source/opt/trim_capabilities_pass.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  // Killing an instruction invalidates the instruction iterator. We must gather
+  // & kill in 2 steps.
+  for (auto instruction : instructions_to_kill) {
+    context()->KillInst(instruction);
+  }
+}
+
+Pass::Status TrimCapabilitiesPass::FilterModuleCapabilities(
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:129]: This remove the extensions as well as the capabilities. It might be good it have three functions instead of two:

```
// One function to remove the OpCapability instructions, and return the capabilities that were removed.
CapabilitySet TrimUnrequiredCapabilities(const CapabilitySet& required_capabilities);

// The same as the existing function.
void TrimExtensionsForCapabilities(required_capabilities, capabilities_to_trim);

// 
Pass::Status TrimCapabilitiesPass::TrimUnrequiredCapabilitiesAndExtensions(const CapabilitySet& required_capabilities) {
  CapabilitySet trimmed_capabilities =  TrimUnrequiredCapabilities(required_capabilities);
  TrimExtensionsForCapabilities(required_capabilities, trimmed_capabilities);
}
```

File: source/opt/trim_capabilities_pass.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  {
+    CapabilitySet missing_capabilities;
+    std::copy_if(
+        required_capabilities.cbegin(), required_capabilities.cend(),
+        std::inserter(missing_capabilities, missing_capabilities.begin()),
+        [](auto capability) {
+          return kSupportedCapabilities.count(capability) == 0;
+        });
+    assert(missing_capabilities.size() == 0 &&
+           "Outdated kSupportedCapabilities.");
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:190]: This would be easier to read as a simple loop:

```
for(spv::Capability c : required_capabilities) {
  assert(kSupportedCapabilities.count(capability) != 0 && "Module is using a capability that is not supported.");
}
```

Three lines vs 9, no lambdas, and no extra data structures.

File: source/opt/trim_capabilities_pass.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+#include "source/opt/pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// This is required for NDK build. The unordered_set/unordered_map
+// implementation don't work with class enums.
+struct UnsignedHash {
+  unsigned operator()(spv::Capability value) const {
+    return static_cast<unsigned>(value);
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.h:34]: What is the point of having a hash table if the value is not hashed? This deserves a comment to explain why you are doing this.

File: source/opt/trim_capabilities_pass.cpp [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  std::vector<Instruction*> instructions_to_kill;
+  for (auto& instruction : context()->extensions()) {
+    assert(instruction.NumOperands() == 1);
+    auto extension = instruction.GetOperand(0).AsString();
+
+    if (required_extensions.count(extension) != 0) {
+      continue;
+    }
+
+    if (trimmed_extensions.count(extension) != 0) {
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:117]: Add a comment explaining why you do not kill an extension even if it is not required by `required_capabilities`.

File: source/opt/trim_capabilities_pass.cpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+#include "source/util/string_utils.h"
+
+namespace spvtools {
+namespace opt {
+
+// Aliases for commonly used types.
+using ExtensionSet = std::unordered_set<std::string>;
+
+// An opcode handler is a function which, given an instruction, returns either
+// the required capability, or spv::Capability::Max. Each handler should
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:36]: Ah misses the comment. Before using optional I always returned an enum, and used Max as "no capability", but changed it to optionnal.

File: source/opt/trim_capabilities_pass.cpp [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+// determine a single capability. If the same instruction could require multiple
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:37]: Sorry need to reword. What I meant is by-design, one handler should only focus on one aspect of a capability to keep them simple and easy to debug/add.

Let's say we have `OpImage` which is changed by capability A and B.
We'd have 2 handlers:
- Handler_OpImage_A
- Handler_OpImage_B
Which checking the condition for their own capability. Maybe A looks only at an operand value while B checks a mix of operand+memory model.

[the remark this answers is not in the document]
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:37]: rephrased.

File: source/opt/trim_capabilities_pass.cpp (hunk 3 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:129]: valid point. Removed the filter function (inlined in the pass function), and added TrimExtensionsForCapabilities.

File: source/opt/trim_capabilities_pass.cpp (hunk 4 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:190]: WAYYYY better, that was bad.

File: source/opt/trim_capabilities_pass.h (hunk 5 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.h:34]: Forgot to call std::hash, the goal here is to provide a hash function for class enums. C++17 handles that, but NDK lags slightly behind on those.
Fixed this code.

File: source/opt/trim_capabilities_pass.cpp (hunk 2 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:90]: Oh, and there is an ExtensionToString! Thanks, will use that instead.

File: source/opt/trim_capabilities_pass.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    spv::Capability::Linkage,
+};
+
+// Those capabilities are never removed from a module because we cannot
+// guess from the SPIR-V only if they are required or not.
+static const CapabilitySet kUntouchableCapabilities{
+    spv::Capability::Shader,
+};
+
+static const std::unordered_map<spv::Capability, std::string, UnsignedHash>
```
[the remark this answers is not in the document]
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.h:62]: Fixed the trivially constructible/destructible part.

For the featureManager, I am not using it because it doesn't remove instructions from the module. The FeatureManager does have a Remove() function, but it only removes it from the set, but doesn't apply the changes to the underlying module.
Probably something we should fix/improve in a refactorization?
Because of this weird thing, I decided to keep a lower-level interaction with instructions directly.

For the extensions/capability mapping, do you mean I should add a function to, given a list of capabilities, get the list of required extensions, and this depending on the spirv version?

If you think this could be useful outside of this pass, sure, we can add that. Subsequent PR or this PR?

**Comment by Keenuts** [comment]
@s-perron as-is the PR includes changes from https://github.com/KhronosGroup/SPIRV-Tools/pull/5344 but the main logic has been rewritten:
 - uses the feature manager
 - uses the grammar
 
 Extensions are a bit more complex than expected, so I had to keep the logic in the pass, as it couldn't be that generic to be useful outside of this.
 
Since we have custom handlers, this initial PR adds support for `StorageInputOutput16` as this is the simplest one I had which doesn't work just with the grammar.
This makes the PR a bit large, if you are OK with a first PR with no custom handlers (but the logic), I can split. But handlers won't be tested.

Once this initial PR is in, the next step would be to increase test coverage for the basic capabilities and enable them, and only then, expand this to more complex capabilities.

File: source/opt/trim_capabilities_pass.cpp [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        continue;
+      }
+
+      const spv_operand_desc_t *desc = {};
+      auto result = context()->grammar().lookupOperand(operand.type, operand.words[0], &desc);
+      if (result != SPV_SUCCESS) {
+        continue;
+      }
+
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities, capabilities);
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp]: Do you need to add extensions here too?

File: source/opt/trim_capabilities_pass.cpp [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+  for (auto it = begin; it != end; it++) {
+    const OpcodeHandler handler = it->second;
+    auto result = handler(instruction);
+    if (result.has_value()) {
+      capabilities->insert(*result);
+    }
+  }
+}
+
+std::pair<CapabilitySet, ExtensionSet> TrimCapabilitiesPass::DetermineRequiredCapabilitiesAndExtensions() const {
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp]: I'd prefer if we had a struct with meaningful names instead of `std::pair<>`. It is hard to know what you are accessing when some just sees `first` and `second` for the uses.

**Review by s-perron — COMMENTED** [review]
This generally looks good. Thanks for putting all of the working into this pass. It will be useful. Let me know when it is in its final form, and I can officially approve it.

File: source/opt/trim_capabilities_pass.cpp (hunk 11 above)
Comment (s-perron) [reply source/opt/trim_capabilities_pass.cpp]: I now see what you did. This is fine.

File: source/opt/trim_capabilities_pass.cpp (hunk 10 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp]: Seems like this comment was done on an old version?

**Comment by Keenuts** [comment]
@s-perron the PR is ready for review 😊

File: source/opt/trim_capabilities_pass.cpp [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+      }
+
+      const spv_operand_desc_t* desc = {};
+      auto result = context()->grammar().lookupOperand(operand.type,
+                                                       operand.words[0], &desc);
+      if (result != SPV_SUCCESS) {
+        continue;
+      }
+
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities,
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp:200]: This is the new location of my earlier comment.

In the first case, you have the code:

```
      extensions->insert(desc->extensions,
                         desc->extensions + desc->numExtensions);
```

but that is not replicated here. I there a reason for that?

File: source/opt/trim_capabilities_pass.cpp (hunk 10 above)
Comment (s-perron) [reply source/opt/trim_capabilities_pass.cpp]: I added a comment where this code was moved. I reviewed the commits separately, which is why it was on a different line.

File: source/opt/trim_capabilities_pass.cpp (hunk 12 above)
Comment (Keenuts) [reply source/opt/trim_capabilities_pass.cpp:200]: There is, AFAIK only one extension enabled by an operand: VulkanMemoryModel. We don't support it for now, so this path is not tested.
To stay coherent, I added it to this loop, but keeping the handler capability-only until we have test cases (so a supported capability requiring this).

**Comment by Keenuts** [comment]
@s-perron ready for another review!
FYI: the second extension insertion is not tested: we don't support any extension requiring it yet.
Same for the handler: it doesn't return extension requirements because it's not required yet.
Adding this would mean adding more support in this initial PR, which makes it even bigger. Seems like we can add this later, as we need it.

File: source/opt/trim_capabilities_pass.cpp [hunk 13]
```diff
@@ -159,46 +159,54 @@ TrimCapabilitiesPass::TrimCapabilitiesPass()
...
     }
 
+    addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities,
+                                  capabilities);
     if (desc->minVersion <= spvVersionForTargetEnv(context()->GetTargetEnv())) {
       extensions->insert(desc->extensions,
                          desc->extensions + desc->numExtensions);
     }
-  }
+  } while (0);
```
Comment (s-perron) [inline source/opt/trim_capabilities_pass.cpp]: It is hard to follow the control flow. Adding the single-iteration loop so you can break is on par with using a goto.  Try

```suggest
  const spv_opcode_desc_t* desc = {};
  auto result =
      context()->grammar().lookupOpcode(instruction->opcode(), &desc);
    if (result == SPV_SUCCESS) {
      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities,
                                  capabilities);
      if (desc->minVersion <= spvVersionForTargetEnv(context()->GetTargetEnv())) {
        extensions->insert(desc->extensions,
                           desc->extensions + desc->numExtensions);
      }
    }
```

**Review by s-perron — CHANGES_REQUESTED** [review]

**Comment by Keenuts** [comment]
Removed the `do/while(0)` block, PR ready.

**Review by s-perron — APPROVED** [review]
